File: src/main/java/au/id/micolous/metrodroid/transit/CardTransitFactory.java
package au.id.micolous.metrodroid.transit;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.Collections;
import java.util.List;

public interface CardTransitFactory<T> {
    @NonNull
    default List<CardInfo> getAllCards() {
        return Collections.emptyList();
    }

    @Nullable
    TransitIdentity parseTransitIdentity(@NonNull T card);

    @Nullable
    TransitData parseTransitData(@NonNull T card);

    boolean check(@NonNull T card);
}

File: src/commonMain/kotlin/au/id/micolous/metrodroid/transit/CardTransitFactory.kt
package au.id.micolous.metrodroid.transit

interface CardTransitFactory<T> {
    val allCards: List<CardInfo>

    fun parseTransitIdentity(card: T): TransitIdentity?

    fun parseTransitData(card: T): TransitData?

    fun check(card: T): Boolean

    val notice: String?
        get() = null
}
